extensions/amp-inputmask/0.1/text-mask.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Mask} from './mask-impl';
+import {OutputMode} from './constants';
+import {Services} from '../../../src/services';
+
+export class TextMask {
+  /**
+   *
+   * @param {!Element} element
+   */
+  constructor(element) {
+    /** @private @const */
+    this.element_ = element;
+
+    /** @private @const */
+    this.document_ = element.ownerDocument;
+
+    /** @private @const */
+    this.outputMode_ = element.getAttribute('mask-output') || OutputMode.RAW;
+
+    const mask = element.getAttribute('mask');
+
+    /** @private @const */
+    this.controller_ = new Mask(element, mask);
+
+    this.controller_.mask();
+
+    Services.formSubmitPromiseForDoc(element).then(formSubmitService => {
+      formSubmitService.beforeSubmit(e => this.handleBeforeSubmit_(e.form));
+    });
+
+    /** @private {?Element} */
+    this.hiddenInput_ = null;
+  }
+
+  /**
+   * Add a hidden input to make the unmasked value available to the form
+   * @param {!HTMLFormElement} form
+   */
+  handleBeforeSubmit_(form) {
+    if (this.outputMode_ == OutputMode.ALPHANUMERIC) {
+      if (!this.hiddenInput_) {
+        const hidden = this.document_.createElement('input');
+        hidden.type = 'hidden';
+        const name = this.element_.name || this.element_.id;
+        hidden.name = `${name}-unmasked`; // TODO(cvializ): detect collisions
```

Comment:
is collision detection needed? IF author used the same name twice , the hidden should follow the same.

I think we should skip adding the hidden if there is no `name` on element.

Reply:
I think collision detection is important so that we don't overwrite someone else's input if their form already has a `${name}-unmasked` input e.g.
```html
<form>
<input name="wow" mask="LLL-LLL" mask-output="alphanumeric">
<input name="wow-masked" type="email"><!-- would get shadowed -->
</form>
```
It's a low risk, but I think it's still something we should try to avoid